test/gtest/uct/ib/test_ud_ds.cc

```diff
@@ -52,21 +52,41 @@ class test_ud_ds : public uct_test {
     }
 
     void test_cep_insert(entity *e, uct_ib_address_t *ib_addr,
-                         uct_ud_iface_addr_t *if_addr, unsigned base);
+                         uct_ud_iface_addr_t *if_addr, unsigned base,
+                         unsigned num_paths);
 
     void create_and_insert_ep(entity *e, uct_ib_address_t *ib_addr,
                               uct_ud_iface_addr_t *if_addr,
-                              uct_ud_ep_conn_sn_t conn_sn, unsigned ep_index,
-                              bool ep_private);
+                              unsigned path_index, uct_ud_ep_conn_sn_t conn_sn,
+                              unsigned ep_index, bool ep_private);
 
     uct_ud_ep_t *get_ep(entity *e, uct_ib_address_t *ib_addr,
-                        uct_ud_iface_addr_t *if_addr,
+                        uct_ud_iface_addr_t *if_addr, unsigned path_index,
                         uct_ud_ep_conn_sn_t conn_sn, bool ep_private);
 
     void insert_ep(entity *e, uct_ib_address_t *ib_addr,
-                   uct_ud_iface_addr_t *if_addr,
+                   uct_ud_iface_addr_t *if_addr, unsigned path_index,
                    uct_ud_ep_conn_sn_t conn_sn, uct_ud_ep_t *ep);
 
+    ucs_conn_sn_t get_conn_sn(uct_ud_iface_t *ud_iface, size_t i,
+                              unsigned num_paths, unsigned path_index)
+    {
+        if (uct_ib_iface_is_roce(&ud_iface->super)) {
+            if (ud_iface->super.config.roce_path_factor == 0) {
+                return i * num_paths + path_index;
+            } else {
+                return i;
+            }
+        } else {
+            unsigned num_path_index_variant =
+                    num_paths / ud_iface->super.path_bits_count;
+            unsigned path_index_variant     =
+                    path_index / ud_iface->super.path_bits_count;
+            return (ucs_conn_sn_t)(i * num_path_index_variant +
+                                   path_index_variant);
```

Comment:
wouldn't it already be part of uct_ud_iface_addr_t ?

Reply:
no, `path_index` is not a part of `uct_ud_iface_addr_t`.
only destination lid is a part of it, which is a result of applying `path_index` to source port or lmc

Reply:
right, if affects lid, so why need it to also affect conn_sn?

Reply:
> if affects lid, so why need it to also affect conn_sn?

if `num_paths % num_path_bits` != 0, so the same lids could be used for the different paths and we should handle this case

Reply:
maybe set IB_LID_PATH_BITS to different value?

Reply:
the values specified in IB_LID_PATH_BITS should be lower than `UCS_BIT(lmc)`

Reply:
I meant this check - https://github.com/openucx/ucx/blob/963b62c6ab73f344d0322f453887ff0d957b364c/src/uct/ib/base/ib_iface.c#L914